File: 17-letter-combinations-of-a-phone-number/solution.py
from collections import deque


class Solution:
  @staticmethod
  def letterComboHelper(digits: str, builder: str) -> list[str]:
    if not digits:
      return [] if not builder else [builder]

    digit_to_letters = {
        '2': 'abc',
        '3': 'def',
        '4': 'ghi',
        '5': 'jkl',
        '6': 'mno',
        '7': 'pqrs',
        '8': 'tuv',
        '9': 'wxyz'
    }

    result: list[str] = []
    for letter in digit_to_letters[digits[0]]:
      result.extend(Solution.letterComboHelper(digits[1:], builder + letter))

    return result

  @staticmethod
  def letterComboIter(digits: str) -> list[str]:
    digit_to_letters = {
        '2': 'abc',
        '3': 'def',
        '4': 'ghi',
        '5': 'jkl',
        '6': 'mno',
        '7': 'pqrs',
        '8': 'tuv',
        '9': 'wxyz'
    }

    result: list[str] = []
    work_deque: deque[tuple[str, str]] = deque([(digits, '')])
    while work_deque:
      curr_digits, builder = work_deque.popleft()
      if not curr_digits:
        if builder:
          result.append(builder)

        continue

      for letter in digit_to_letters[curr_digits[0]]:
        work_deque.append((curr_digits[1:], builder + letter))

    return result

  def letterCombinations(self, digits: str) -> list[str]:
    # return Solution.letterComboHelper(digits, '')
    return Solution.letterComboIter(digits)

File: 17-letter-combinations-of-a-phone-number/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>

class Solution
{
private:
  static std::vector<std::string> letterComboHelper(const std::string &digits, std::string builder)
  {
    if (digits.empty())
    {
      if (builder.empty())
      {
        return {};
      }

      return {builder};
    }

    const std::unordered_map<char, std::string> digit_to_letters{{'2', "abc"}, {'3', "def"}, {'4', "ghi"}, {'5', "jkl"}, {'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}};

    const auto iter = digit_to_letters.find(digits[0]);
    if (iter == digit_to_letters.end())
    {
      return {};
    }

    std::vector<std::string> result;
    for (const char letter : iter->second)
    {
      for (auto word : Solution::letterComboHelper(digits.substr(1), builder + letter))
      {
        result.emplace_back(word);
      }
    }

    return result;
  }

public:
  std::vector<std::string> letterCombinations(std::string digits)
  {
    return Solution::letterComboHelper(digits, "");
  }
};